snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java

```diff
@@ -169,12 +171,16 @@ private boolean isSnapshotIdInvalid(final String snapshotId) {
   }
 
   private boolean writeReceivedSnapshotChunk(
-      final SnapshotChunk snapshotChunk, final Path snapshotFile) throws IOException {
-    Files.write(
-        snapshotFile,
-        snapshotChunk.getContent(),
-        StandardOpenOption.CREATE_NEW,
-        StandardOpenOption.WRITE);
+      final SnapshotChunk snapshotChunk, final Path snapshotFile) {
+    try (var file = new RandomAccessFile(snapshotFile.toFile(), "ws")) {
+      final FileChannel channel = file.getChannel();
+      final ByteBuffer buffer = ByteBuffer.wrap(snapshotChunk.getContent());
+      channel.write(buffer);
+    } catch (IOException e) {
+      LOGGER.warn("Failed to write snapshot chunk {}", snapshotChunk, e);
+      return FAILED;
+    }
```

Comment:
You can open a file channel directly by using `FileChannel.open`.

Additionally, the way `FileChannel.write` works, you want to write incrementally in small buffers. If you look into it, it uses a direct buffer to write all the time. This is a good indicator that we should eventually use direct buffers here, but for now as we can't/don't, it will use a cached native buffer - unless the buffer is too big, at which point it allocates it fresh every time. It'll still have to copy the memory though, but that's not really any different than with the old I/O since it still had to cross the JVM boundary before being written.

Additionally, you have to keep in mind that `FileChannel#write` may not write all bytes - so you want to make sure it does by running:

```java
while (buffer.hasRemaining()) {
  final var chunkSize = Math.min(buffer.capacity(), buffer.position() + IoUtil.BLOCK_SIZE);
  channel.write(buffer.limit(chunkSize));
}
```

I use `IoUtil.BLOCK_SIZE` here, but tbh I'm not sure what's the real optimal size. Generally aligned sizes are just better (since the underlying buffer is aligned as well, even if you don't use an aligned size), but I know some people report using bigger blocks will work better - might be true especially since we use network attached storage, so we might want to use something more like 512KB. I think over 1MB it might not get cached anyway, and 512KB is a good size for our snapshot files.

Reply:
Why is using more and "smaller" writes better? Looking at `write()`'s implementation, the maximum buffer (until it's too big to cache) is Long.MAX_VALUE unless 'jdk.nio.maxCachedBufferSize' is set. I googled around briefly for some best practices/analysis on this but didn't find anything conclusive
Also, that code snippet limits the buffer size and then calls `hasRemaining()` but I think we need to reset the limit to the capacity. Otherwise, if a buffer has more than BLOCK_SIZE, we would write that and then exit since `hasRemaining()` would return false

Reply:
Yeah, that block is wrong, I just wrote it otoh, you're right that just setting the limit wouldn't work. I looked around, and I also couldn't find conclusive evidence that writing in smaller chunks is "better" - this is something I learned as a best practice a long time ago, and rule of thumb was always to read/write in smaller, aligned chunks. What I did find seems to corroborate this, but I couldn't find anyone giving a conclusive reason, just a bunch of benchmarks.

Regarding the buffer size, one potential issue I see here is allocating and caching an unknown buffer size. Keep in mind that direct memory is not managed by the heap, and while it's garbage collected, it works a bit differently (in that you don't really know when the native memory is really reclaimed). Additionally, it uses a different pool with different configuration (e.g. `-XX:MaxDirectMemorySize`). It's probably usually fine, but IDK - I'm not too hot about allocating a randomly sized buffer, even when it's cached. I'm not sure when is that cache cleaned - when the thread exits? But if it runs on the actor thread, then it's never cleaned? I'd rather get a smaller buffer if it's going to be long living.

IDK - wdyt?

Reply:
Thanks, I just wanted to see if there was some mechanism I wasn't getting, performance/buffer issues are not very transparent. I definitely understand the concern of not allocating a buffer of unknown size. As it is atm, the buffer will be at most 512Kb so it should meet those criteria